service/src/main/java/com/epam/esm/service/impl/GiftCertificateServiceImpl.java

```diff
@@ -0,0 +1,454 @@
+     * @param certificateId  id of updatable GiftCertificate entity.
+     * @return GiftCertificateDto with id and tags.
+     * @throws GiftCertificateNotFoundException if GiftCertificate entity with given id doesn't exist in db.
+     * @since 1.0
+     */
+    @Override
+    @Transactional(isolation = Isolation.READ_COMMITTED)
+    public GiftCertificateDto update(GiftCertificateDto certificateDto, long certificateId) {
+
+        Optional<GiftCertificate> certificateOpt = giftCertificateDao.getById(certificateId);
+
+        certificateOpt.orElseThrow(() ->
+                new GiftCertificateNotFoundException(String.format("GiftCertificate with id: %d doesn't exist in DB", certificateId)));
+
+        String currentTime = DateProvider.getNowAsString();
+        certificateDto.setLastUpdateDate(currentTime);
+
+        GiftCertificate giftCertificate = modelMapper.map(certificateDto, GiftCertificate.class);
+        giftCertificateDao.update(giftCertificate, certificateId);
+
+        certificateTagsDao.deleteAllTagsForCertificate(certificateId);
+
+        List<TagDto> tags = certificateDto.getTags();
+        if (tags != null) {
+            tags.forEach(tagDto -> {
+                Tag tag = addTagToCertificate(tagDto, certificateId);
+                tagDto.setId(tag.getId());
+            });
+        }
+
+        certificateDto.setId(certificateId);
+        certificateDto.setCreateDate(certificateOpt.get().getCreateDate());
+        return certificateDto;
+
+    }
+
+
+    /**
+     * This method deletes GiftCertificate entity with given id from db.
+     *
+     * @param id id of deletable GiftCertificate entity.
+     * @throws GiftCertificateNotFoundException if GiftCertificate entity with given id doesn't exist in db.
+     * @since 1.0
+     */
+    @Override
+    public void delete(long id) {
+        boolean isDeleted = giftCertificateDao.delete(id);
+        if (!isDeleted) {
+            throw new GiftCertificateNotFoundException(String.format("GiftCertificate with id: %d doesn't exist in DB", id));
+        }
+    }
+
+    /**
+     * This method partly updates GiftCertificate entity.Also the method link passed tags with the GiftCertificate
+     * entity and, if the tag doesn't exist in db, creates it.
+     *
+     * @param giftCertificatePatchDto DTO contains field for partial updating GiftCertificate entity.
+     * @param certificateId           id of updatable GiftCertificate entity.
+     * @return updated GiftCertificateDto with id and tags.
+     * @throws GiftCertificateNotFoundException if GiftCertificate entity with given id doesn't exist in db.
+     * @since 1.0
+     */
+    @Override
+    @Transactional(isolation = Isolation.READ_COMMITTED)
+    public GiftCertificateDto patch(GiftCertificatePatchDto giftCertificatePatchDto, long certificateId) {
+        Optional<GiftCertificate> certificateOpt = giftCertificateDao.getById(certificateId);
```

Comment:
To much for me) please think about splitting and simplifying

Reply:
I did a little refactor and added some comments
commit 79a44a5